### PR TITLE
Container native provenance predicate

### PR DESCRIPTION
**Pull request by priyawadhwa**
Creates a new `tekton-provenance` format type, which uses the provenance defined in `pkg/chains/provenance`

### CODE REVIEW

File: PROVENANCE_SPEC.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+This should be set if the build was created from a template or other type of file defined out of the build request itself.
+
+For a TaskRun, this could be a PipelineRun, an OCI image, or reference a Task in-cluster, with the following prefixes:
+
+* `oci://`
+* `task://`
+* `clusterTask://`
+* `pipeline://`
+
+`recipe`, optional, object
```
Comment (dlorenc) [inline PROVENANCE_SPEC.md:91]: I think we want to make recipe itself mandatory.

File: PROVENANCE_SPEC.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        "digest": null
+      }
+    ]
+  }
+}
+```
+
+## Fields
+
+`subject`, required, `object`
```
Comment (dlorenc) [inline PROVENANCE_SPEC.md:61]: I think we decided we might want to make subject repeated. Still not entirely sure...

File: PROVENANCE_SPEC.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+Pins down the material to an immutable version (e.g. pins down a Git repo to a specific commit)
+TODO
+
+
+## Example
+
+```json
+[
+  {
+    "_type": "",
```
Comment (dlorenc) [inline PROVENANCE_SPEC.md:128]: We also forgot a place for builder.id and other build metadata (version, etc.)

File: pkg/chains/signing.go [hunk 4]
```diff
@@ -223,6 +223,8 @@ func (ts *TaskRunSigner) SignTaskRun(ctx context.Context, tr *v1beta1.TaskRun) e
 				ts.Logger.Warnf("Unable to marshal payload: %v", signerType, obj)
 				continue
 			}
+			fmt.Println("~~~~~~~~~~~~ signing payload: ~~~~~~~")
```
Comment (dlorenc) [inline pkg/chains/signing.go:226]: you mean `fmt.Println("HERE")` :)

Comment (priyawadhwa) [reply pkg/chains/signing.go:226]: hahahaha

File: PROVENANCE_SPEC.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    },
+    "predicate": {
+      "invocation": {
+        "parameters": [
+          "PROJECT_ID=priya-wadhwa",
+          "BAZEL_BUILD_ARGS=nodejs14_amd64,nodejs14_arm64"
+        ],
+        "recipe_uri": "task://distroless-build-from-git",
+        "event_id": "10f420eb-3b64-4625-84ab-c23da3eb6207",
+        "id": "tekton"
```
Comment (priyawadhwa) [inline PROVENANCE_SPEC.md:148]: change to builder.id

File: pkg/chains/formats/intotoite6/intotoite6.go [hunk 6]
```diff
@@ -57,93 +59,192 @@ func (i *InTotoIte6) CreatePayload(obj interface{}) (interface{}, error) {
...
 			continue
 		}
+		m := provenance.ProvenanceMaterial{
+			Digest: provenance.DigestSet{},
+		}
+		for _, param := range input.ResourceSpec.Params {
+			if param.Name == "url" {
+				m.URI = param.Value
+			}
+			if param.Name == "revision" {
```
Comment (priyawadhwa) [inline pkg/chains/formats/intotoite6/intotoite6.go:126]: find the exact commit from tr.status.resourceresults & join on the name

File: pkg/chains/formats/intotoite6/intotoite6.go [hunk 7]
```diff
@@ -57,93 +59,192 @@ func (i *InTotoIte6) CreatePayload(obj interface{}) (interface{}, error) {
...
-			definedInMaterial = intP(i)
-			break
+
+	// get parameters
+	var params []string
+	for _, p := range tr.Spec.Params {
+		params = append(params, fmt.Sprintf("%s=%v", p.Name, p.Value))
+	}
+	// add params with default values
+	if ts := tr.Spec.TaskSpec; ts != nil {
```
Comment (priyawadhwa) [inline pkg/chains/formats/intotoite6/intotoite6.go:148]: switch to status

File: pkg/chains/formats/intotoite6/intotoite6.go [hunk 8]
```diff
@@ -57,93 +59,192 @@ func (i *InTotoIte6) CreatePayload(obj interface{}) (interface{}, error) {
...
-	// find the index of the Git material, if it exists, in the newly sorted list
-	var definedInMaterial *int
-	for i, u := range m {
-		if u.URI == gitURL {
-			definedInMaterial = intP(i)
-			break
+
+	// get parameters
+	var params []string
+	for _, p := range tr.Spec.Params {
```
Comment (priyawadhwa) [inline pkg/chains/formats/intotoite6/intotoite6.go:144]: switch to status

File: pkg/chains/formats/intotoite6/intotoite6.go (hunk 7 above)
Comment (priyawadhwa) [reply pkg/chains/formats/intotoite6/intotoite6.go:148]: remove

File: pkg/chains/formats/intotoite6/intotoite6.go [hunk 9]
```diff
@@ -183,53 +284,68 @@ func gitInfo(tr *v1beta1.TaskRun) (commit string, url string) {
...
+	var subjects []provenance.Subject
+	type imageDetails struct {
+		url    string
+		digest string
+	}
+	images := map[string]*imageDetails{}
+
+	// look for results IMAGE_URL and IMAGE_DIGEST
+	for _, res := range tr.Status.TaskRunResults {
+		if strings.HasSuffix(res.Name, "IMAGE_URL") {
```
Comment (priyawadhwa) [inline pkg/chains/formats/intotoite6/intotoite6.go:297]: this shouldn't require a prefix

File: pkg/chains/provenance/provenance.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+// ProvenanceMetadata contains metadata for the built artifact.
+type ProvenanceMetadata struct {
+	// Use pointer to make sure that the abscense of a time is not
+	// encoded as the Epoch time.
+	BuildStartedOn  *time.Time `json:"buildStartedOn,omitempty"`
+	BuildFinishedOn *time.Time `json:"buildFinishedOn,omitempty"`
+	// removed: Completeness
+	Reproducible bool `json:"reproducible"`
```
Comment (priyawadhwa) [inline pkg/chains/provenance/provenance.go:86]: use an annotation

File: pkg/chains/formats/provenance/provenance.go [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+		}
+	}
+	return m
+}
+
+// add any Git specification to materials
+func materials(tr *v1beta1.TaskRun) []provenance.ProvenanceMaterial {
+	var mats []provenance.ProvenanceMaterial
+
+	gitCommit, gitURL := gitInfo(tr)
```
Comment (priyawadhwa) [inline pkg/chains/formats/provenance/provenance.go:125]: only do this if there aren't any resources

File: pkg/chains/provenance/provenance.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+	// added here
+	Name          string `json:"name"`
+	PredicateType string `json:"predicateType"`
+	// removed: option for multiple subjects
+	Subject Subject `json:"subject"`
+}
+
+// ProvenanceStatement is the definition for an entire provenance statement.
+type ProvenanceStatement struct {
```
Comment (priyawadhwa) [inline pkg/chains/provenance/provenance.go:42]: TektonProvenanceStatement

File: pkg/chains/formats/provenance/provenance.go [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+				return s
+			}
+		}
+	}
+
+	return v1beta1.Step{}
+}
+
+func (i *Provenance) Type() formats.PayloadType {
+	return formats.PayloadTypeProvenance
```
Comment (priyawadhwa) [inline pkg/chains/formats/provenance/provenance.go:277]: rename

**Comment by dlorenc** [comment]
LGTM! Want to squash?

**Pull request merged**